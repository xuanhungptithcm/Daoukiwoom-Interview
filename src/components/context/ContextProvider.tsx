import { PropsWithChildren, ReactElement, useState } from "react";
import { IReactContext } from "./types";

interface IContextProviderProps<T> {
  Context: IReactContext<T>;
  initialState: T;
}

function ContextProvider<T>({
  Context,
  initialState,
  children,
}: PropsWithChildren<IContextProviderProps<T>>): ReactElement {
  const state = useState<T>(initialState);

  return <Context.Provider value={state}>{children}</Context.Provider>;
}

export default ContextProvider;
