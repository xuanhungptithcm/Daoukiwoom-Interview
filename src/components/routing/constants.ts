export enum AUTH_PATH {
  LOGIN_PAGE = "/login",
  AUTH_PAGE = "/auth",
}
