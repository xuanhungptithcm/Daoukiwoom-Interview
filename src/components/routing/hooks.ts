import { useEffect, useMemo, useState } from "react";

const useIsUserLoggedIn = () => {
  return true;
}
export default useIsUserLoggedIn;