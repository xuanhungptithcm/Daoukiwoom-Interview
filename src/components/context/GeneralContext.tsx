import { PropsWithChildren, ReactElement } from "react";
import ApiHandler from "../../api/ApiHandler";

import { GeneralDataContext, initialGeneralData } from "./constants";
import ContextProvider from "./ContextProvider";
import { IGenericProps } from "./types";

const GeneralContext = ({
  children,
}: PropsWithChildren<IGenericProps>): ReactElement => {
  return (
    <ContextProvider
      initialState={initialGeneralData}
      Context={GeneralDataContext}
    >
      <ApiHandler />
      {children}
    </ContextProvider>
  );
};

export default GeneralContext;
