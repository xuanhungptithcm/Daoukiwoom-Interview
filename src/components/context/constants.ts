import React from "react";
import { makeContext } from "./helpers";

import { IGeneralData } from "./types";

export const initialSetter: React.Dispatch<React.SetStateAction<unknown>> =
  // this is just a placeholder function, it will never be used
  () => {
    // EMPTY
  };

export const initialGeneralData: IGeneralData = {
  isLoggedIn: true,
  userInfo: {
    createdAt: "",
    name: "",
    avatar: "",
    username: "",
    password: "",
    id: "",
  },
};

export const GeneralDataContext = makeContext(initialGeneralData);
