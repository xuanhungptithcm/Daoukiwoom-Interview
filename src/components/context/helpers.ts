import { createContext } from "react";

import { initialSetter } from "./constants";
import { IContextState } from "./types";

export const makeContext = <T>(initial: T) => {
  return createContext<IContextState<T>>([initial, initialSetter]);
};