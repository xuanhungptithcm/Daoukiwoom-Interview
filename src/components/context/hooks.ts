import { useContext, useMemo } from "react";
import { GeneralDataContext } from "./constants";

export const useGeneralContext = () => {
  const [{ isLoggedIn, userInfo }, setGeneralContext] =
    useContext(GeneralDataContext);

  return {
    isLoggedIn,
    userInfo,
    setGeneralContext,
  };
};

export const useIsUserLogIn = () => {
  const { isLoggedIn } = useGeneralContext();

  return useMemo(() => isLoggedIn, [isLoggedIn]);
};

export const useUserInfo = () => {
  const { userInfo } = useGeneralContext();
  return useMemo(() => userInfo, [userInfo]);
};
