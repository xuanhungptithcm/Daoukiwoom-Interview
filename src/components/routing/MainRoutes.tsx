import { ReactElement } from "react";
import { BrowserRouter, Redirect, Route, Switch } from "react-router-dom";
import LoginPage from "../../pages/auth/login/LoginPage";
import ErrorPage from "../../pages/error/ErrorPage";
import BasePage from "../base/BasePage";
import { useIsUserLogIn } from "../context/hooks";
import { AUTH_PATH } from "./constants";

const Routes = (): ReactElement => {
  const isLogged = useIsUserLogIn();
  console.log("ðŸš€ ~ file: MainRoutes.tsx ~ line 11 ~ Routes ~ isLogged", isLogged)
  
  return (
    <BrowserRouter basename="">
      <Switch>
        <Route path={AUTH_PATH.NOT_FOUND} component={ErrorPage} />
        <Route path={AUTH_PATH.LOGIN_PAGE} component={LoginPage} />
        {!isLogged && (
          <Route>
            <Redirect to={AUTH_PATH.LOGIN_PAGE} />
          </Route>
        )}
        <BasePage isLogged={isLogged} />
        <Redirect from="*" to={AUTH_PATH.NOT_FOUND} />
      </Switch>
    </BrowserRouter>
  );
};

export default Routes;
