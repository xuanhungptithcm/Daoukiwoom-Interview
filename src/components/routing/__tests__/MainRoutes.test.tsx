/**
 * @jest-environment jsdom
 */

import { shallow } from "enzyme";
import { useIsUserLogIn } from "../../context/hooks";
import { AUTH_PATH } from "../constants";
import MainRoutes from "../MainRoutes";

describe("MainRoutes", () => {
  beforeEach(() => {
    jest.mock("uid", () => () => 1);
    jest.mock("../../../routes/routes", () => 1);
  });
  test("should properly render with authorized", () => {
    const routes = [
      {
        key: "1",
        path: AUTH_PATH.LOGIN_PAGE,
        component: "B",
        requiredAuthen: false,
        exact: true,
        isAuthorized: false,
      },
      {
        key: "2",
        path: "/",
        component: "B",
        requiredAuthen: false,
        exact: true,
        isAuthorized: false,
      },
    ];
    jest.mock("../../../routes/routes", () => routes);

    (useIsUserLogIn as any) = () => true;
    expect(shallow(<MainRoutes />)).toMatchSnapshot();
  });
  test("should properly render with no authorized", () => {
    const routes = [
      {
        key: "1",
        path: AUTH_PATH.LOGIN_PAGE,
        component: "B",
        requiredAuthen: false,
        exact: true,
        isAuthorized: false,
      },
      {
        key: "2",
        path: "/",
        component: "B",
        requiredAuthen: false,
        exact: true,
        isAuthorized: false,
      },
    ];
    jest.mock("../../../routes/routes", () => routes);

    jest.mock("../../context/hooks", () => ({
      useIsUserLogIn: () => false,
    }));
    (useIsUserLogIn as any) = () => false;
    expect(shallow(<MainRoutes />)).toMatchSnapshot();
  });
});
