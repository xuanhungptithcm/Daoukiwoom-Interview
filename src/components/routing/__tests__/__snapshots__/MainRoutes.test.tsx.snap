// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainRoutes should properly render with authorized 1`] = `
<Switch>
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={true}
    isAuthorized={true}
    key="b76e1a7c16e"
    path="/login"
    requiredAuthen={false}
  />
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={true}
    isAuthorized={true}
    key="6e1a7c16e66"
    path="/signup"
    requiredAuthen={false}
  />
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={true}
    isAuthorized={true}
    key="76e1a7c16e6"
    path=""
    requiredAuthen={true}
  />
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={true}
    isAuthorized={true}
    key="e1a7c16e667"
    path="/dashboard"
    requiredAuthen={true}
  />
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={false}
    isAuthorized={true}
    key="1a7c16e6678"
    path="*"
    requiredAuthen={false}
  />
</Switch>
`;

exports[`MainRoutes should properly render with no authorized 1`] = `
<Switch>
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={true}
    isAuthorized={false}
    key="b76e1a7c16e"
    path="/login"
    requiredAuthen={false}
  />
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={true}
    isAuthorized={false}
    key="6e1a7c16e66"
    path="/signup"
    requiredAuthen={false}
  />
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={true}
    isAuthorized={false}
    key="76e1a7c16e6"
    path=""
    requiredAuthen={true}
  />
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={true}
    isAuthorized={false}
    key="e1a7c16e667"
    path="/dashboard"
    requiredAuthen={true}
  />
  <RouteWithSubRoutes
    component={
      Object {
        "$$typeof": Symbol(react.lazy),
        "_init": [Function],
        "_payload": Object {
          "_result": [Function],
          "_status": -1,
        },
      }
    }
    exact={false}
    isAuthorized={false}
    key="1a7c16e6678"
    path="*"
    requiredAuthen={false}
  />
</Switch>
`;
